executor/partition_table_test.go

```diff
@@ -305,6 +305,82 @@ func (s *partitionTableSuite) TestGlobalStatsAndSQLBinding(c *C) {
 	tk.MustIndexLookup("select * from tlist where a<1")
 }
 
+func (s *partitionTableSuite) TestDirectReadingwithJoin(c *C) {
+	if israce.RaceEnabled {
+		c.Skip("exhaustive types test, skip race test")
+	}
+
+	tk := testkit.NewTestKitWithInit(c, s.store)
+	tk.MustExec("create database test_dr_join")
+	tk.MustExec("use test_dr_join")
+	tk.MustExec("set @@tidb_partition_prune_mode = 'dynamic'")
+
+	// hash and range partition
+	tk.MustExec("create table thash (a int, b int, key(a)) partition by hash(a) partitions 4;")
+	tk.MustExec(`create table trange (a int, b int, key(a)) partition by range(a) (
+		  partition p0 values less than(1000),
+		  partition p1 values less than(2000),
+		  partition p2 values less than(3000),
+		  partition p3 values less than(4000));`)
+
+	// regualr table
+	tk.MustExec(`create table tnormal (a int, b int, key(a));`)
+	tk.MustExec(`create table touter (a int, b int);`)
+
+	// generate some random data to be inserted
+	vals := make([]string, 0, 2000)
+	for i := 0; i < 2000; i++ {
+		vals = append(vals, fmt.Sprintf("(%v, %v)", rand.Intn(4000), rand.Intn(4000)))
+	}
+	tk.MustExec("insert into trange values " + strings.Join(vals, ","))
+	tk.MustExec("insert into thash values " + strings.Join(vals, ","))
+	tk.MustExec("insert into tnormal values " + strings.Join(vals, ","))
+	tk.MustExec("insert into touter values " + strings.Join(vals, ","))
+
+	// test indexLookUp + hash
+	// explain select /*+ INL_JOIN(touter, tinner) */ * from touter join tinner use index(a) on touter.a = tinner.a;
+	queryPartition := fmt.Sprintf("select /*+ INL_JOIN(touter, thash) */ * from touter join thash use index(a) on touter.a = thash.a")
```

Comment:
How about adding some special selection conditions and testing varied join types:
```
on touter.a = thash.a and b > ?;
on touter.a != thash.a and b > ?;
on touter.a < thash.a and b > ?;
```

Reply:
Ok.

As for varied join types, do you mean types of outer join and inner join, or types of hash join and index look up join?

Reply:
Yes,  outer/inner/semi/... join.

Reply:
If this test only intends to test IndexJoin with different direct reading functors as its inner side, just testing `inner` join is OK since IndexJoin cannot support other types of join.